### PR TITLE
APNG support

### PR DESCRIPTION
**Pull request by Poker-sang**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Implement APNG decoding and encoding based on PNG

### CODE REVIEW

File: src/ImageSharp/Formats/Png/APngDisposeOperation.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/APngDisposeOperation.cs:9]: `PngDisposalMethod` to normalize with Gif.

File: src/ImageSharp/Formats/Png/APngFrameMetadata.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/APngFrameMetadata.cs:11]: `PngFrameMetadata`

File: src/ImageSharp/Formats/Png/Chunks/APngAnimationControl.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/APngAnimationControl.cs:8]: This reads like it should be a readonly struct.

File: src/ImageSharp/Formats/Png/Chunks/APngAnimationControl.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/APngAnimationControl.cs:22]: `RepeatCount` same as Gif.

File: src/ImageSharp/Formats/Png/Chunks/APngFrameControl.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/APngFrameControl.cs:109]: Can we move this to a throwhelper?

File: src/ImageSharp/Formats/Png/Chunks/PngPhysical.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/PngPhysical.cs:13]: Not sure why you renamed this.

File: src/ImageSharp/Formats/Png/Chunks/APngFrameControl.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/APngFrameControl.cs:12]: `FrameControl` is fine.

File: src/ImageSharp/Formats/Png/Chunks/APngAnimationControl.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/APngAnimationControl.cs:8]: `AnimationControl` is fine, we don't expose the type and it's namespaced.

File: src/ImageSharp/Formats/Png/MetadataExtensions.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/MetadataExtensions.cs:27]: You can drop the `A` prefix.

File: src/ImageSharp/Formats/Png/PngConstants.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngConstants.cs:83]: I absolutely love that we will be able to do this but I don't think we should change the language version to allow it.

File: src/ImageSharp/Formats/Png/PngEncoder.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngEncoder.cs:25]: No need for this. The encoder should automatically encode multiple frames as animated png.

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngEncoderCore.cs:28]: I guess...

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:34]: I don't think you need this field.

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:606]: Missing parameter

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:332]: When decoding multi-frame images the decoder should respect the `DecoderOptions.MaxFrames` property.

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:800]: Not sure why you reversed this. Less intuitive when incrementing.

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:333]: I don't think we need the additional checks here. A standard png will not have this chunk.

Do you have examples of APNGs with multiple animation controls?

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:468]: I can't stand this syntax. Pattern matching should be used to simplify readability, not obfuscate it.

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:720]: All frames within a decoded image must be the same size. The decoded data should be offset against the frame. See the gif decoder.

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:216]: This feels very messy. I think it needs reworking. Separate methods are fine.

File: src/ImageSharp/ImageSharp.csproj
Comment (JimBobSquarePants) [inline src/ImageSharp/ImageSharp.csproj:16]: We don't want to change this just now. It limits contribution.

**Review by JimBobSquarePants — CHANGES_REQUESTED** [review]
Thanks for this, I really appreciate your efforts!

 I've tried to offer guidance as to what need to change. Please let me know if anything doesn't make sense.

File: src/ImageSharp/Formats/Png/Chunks/PngPhysical.cs
Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/Chunks/PngPhysical.cs:13]: I was trying to make all trunks type under the same naming style. Says PngHeader, PngTextData, **PngPhysical**, APngAnimationControl, APngFrameControl. Maybe I should rename them to Header, TextData, Physical, AnimationControl, FrameControl?

**Comment by Poker-sang** [comment]
Thanks for your review! It is rigorous and easy for me to understand. I'm sorry for accidentally formatting some code that confused you (e.g. x++ to ++x). Now it's all reverted.

File: src/ImageSharp/Formats/Png/PngScanlineProcessor.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngScanlineProcessor.cs:312]: Why not slice both spans? Your scanline can only be <= to header width.

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:674]: Pass `FrameControl` via `in` to avoid copying.

Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/PngDecoderCore.cs:674]: Do I need to put `in` before all `FrameControl`'s? Same as here.

File: src/ImageSharp/Formats/Png/PngEncoder.cs
Comment (antonfirsov) [inline src/ImageSharp/Formats/Png/PngEncoder.cs:20]: If `QuantizingImageEncoder.Quantizer == null` is a valid state, shouldn't it be nullable?

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs
Comment (antonfirsov) [inline src/ImageSharp/Formats/Png/PngEncoderCore.cs:103]: Shouldn't this field be nullable instead?

File: src/ImageSharp/Formats/Png/PngEncoder.cs
Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/PngEncoder.cs:20]: hmm yes, but `QuantizingImageEncoder.Quantizer` is also used in encoders in many other formats. In order not to change more code, I suppressed the nullable notification here.

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs
Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/PngEncoderCore.cs:103]: Most of the usages of these fields are not null. It is possible to be null only when dispose. So I also used null-forgiving here for the sake of code brevity.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngEncoderCore.cs:103]: You can remove the explicit set to null then from disposal. GC will clean up.

File: src/ImageSharp/Formats/Png/PngEncoder.cs
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngEncoder.cs:20]: It shouldn't be nullable anywhere really, however we're unable to lazily set the quantizer to the required bitdepth without it being nullable here. Best just leaving nullability checks disabled in this file for now. The core encoder deals with nullability. 

I want to have a good look at the quantizers again for V4.

Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/PngEncoder.cs:20]: Then we can disable nullable in the `QuantizingImageEncoder.cs`. This allows us to use nullable in the `Encoder` and avoid property `Quantizer`'s null checking.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngEncoder.cs:20]: You can't do that because that breaks the ability to use a palette quantizer in the presence of a color palette in the metadata #2455  If someone explicitly sets a quantizer they are showing intent.

Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/PngEncoder.cs:20]: I am sorry. I will revert the enable disable in `PngEncoder.cs`.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngEncoder.cs:20]: Nothing to apologize for! 🙂

**Comment by JimBobSquarePants** [comment]
@Poker-sang Sorry this is taking so long to merge. There's a bit of a backlog we need to shift first to ensure there are no functional conflicts.

**Comment by Poker-sang** [comment]
It's all right. Take your time please.

**Comment by JimBobSquarePants** [comment]
@Poker-sang Thanks again for your patience. Following the merge of #2485 there's a few conflicts here. Would you be able to update your PR to include the updates from main? If not, I can have a try.

**Comment by Poker-sang** [comment]
I can have a try. But I'm sorry I can't handle it for I'm busy these days. If you're not in a hurry you can wait a few days for me. ☺️ Or you can handle it the way you want if it doesn't bother you.

**Comment by JimBobSquarePants** [comment]
I'm having a go at merging main back into this PR now.

**Comment by JimBobSquarePants** [comment]
@Poker-sang I've merged the code plus performed some cleanup. We're getting pretty close now, I think. 

Things to note:
- I removed the offsets/widths from the frame metadata. When decoding we should decode to an offset within the full frame and the metadata values become meaningless as soon as we transform the image. We should, in the future, be able to optimize the encoding process by clipping the frames to their non-transparent bounds and encoding only the remaining data with appropriate control offsets (not required for this PR).
- I optimized the code somewhat, we were allocating far too much when cloning and writing the frame data.
- I've renamed some things in the metadata and will have a thinking about renaming others to match gif naming convention. I'm also thinking we should use our Rational type for the delay.

It doesn't seem to be that we are using the blending information at all currently during decode. We'll have to implement that, maybe gif can be used as a reference (I imagine the approach is very similar). 

We definitely need more samples and comprehensive tests. 

There's a few here.
http://littlesvr.ca/apng/samples.html

But maybe we can use this to make more.
https://sourceforge.net/projects/apngasm/

File: src/ImageSharp/Formats/Png/Chunks/AnimationControl.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/Chunks/AnimationControl.cs:26]: RepeatCount

**Comment by JimBobSquarePants** [comment]
> Thank you so much for helping me resolve the conflict. Your code is very clear and impressive! Regarding the points you mentioned, I think I can explain my previous considerations:
> 
> * You removed the width and offset fields, but I'm concerned that this will result in images that are unclipped and large.
> * The naming and type of the original field was something I just wanted to keep in line with the standard documentation. Therefore, I thought it would be fine to change it to some other more understandable name. I also think it is necessary to change the type of the field `Delay` to `Rational` and to be consistent with the GIF.
> * We do need more APNG test samples. I previously referenced the images in https://philip.html5.org/tests/apng/tests.html. Not all of them are available to us, but I think some will help.
> * And since several formats of animated images are implemented. Maybe the type `ImageFrame` needs some slight refactoring to facilitate the handling of animated images.

Happy to help! 

To answer your questions.

The reason they were removed is that they're useful only for placing the data correctly on decode. 
During decode (for gif) we use those offsets to place pixel data into the frame buffer. You can end up with something like this.

```
+------------------------------------------+
|                                          |
|                                          |
|                                          |
|     x,y                                  |
|      +---------------------------+       |
|      |                           |       |
|      |                           |       |
|      |                           |       |
|      |                           |       |
|      |                           | h     |
|      |                           |       |
|      |                           |       |
|      |                           |       |
|      |                           |       |
|      +---------------------------+       |
|                   w                      |
|                                          |
|                                          |
|                                          |
|                                          |
+------------------------------------------+
```

Any transform applied to that image (resize, skew, rotate, etc) will adjust those positions and we have no mechanism by which we can ensure they can be updated accurately. Imagine if a 3rd party wrote a general warping algorithm, there's simply no way to keep track. In addition, any drawing operations could overflow those original bounds. 

The only way to handle this is to keep all frames the same size and place the pixel data in the buffer on decode, we then discard the metadata and, if possible, recalculate it on encode by clipping the quantized frame buffer based on the presence of transparent pixels. You can see that in action in our gif encoder.

https://github.com/SixLabors/ImageSharp/blob/d93bc6cca2383095c2a08db85861b88a1c94c43c/src/ImageSharp/Formats/Gif/GifEncoderCore.cs#L403

Now hopefully we can introduce a general-purpose quantizing utility which will allow us to apply the same approach to all quantized buffers, but I need to have a think about that as we need to cater for multple transparent palette entry support in a performant manner. For the time being it's best to simply encode the whole buffer.

Naming is hard 😄 We tend to use the specification names for internal types but always attempt to normalize property names for public types for API consistency.

I'm curious to hear what changes you deem required for `ImageFrame` as I don't see anything currently that is an additional requirement for animated formats though am always up for good ideas.

**Comment by Poker-sang** [comment]
hmmm say, almost all animated image formats. Frames have a `delay` property. Maybe we can give the `Delay` as a property to the type `ImageFrame` instead of using metadata?

**Comment by JimBobSquarePants** [comment]
> > I'm curious to hear what changes you deem required for `ImageFrame` as I don't see anything currently that is an additional requirement for animated formats though am always up for good ideas.
> 
> hmmm say, almost all animated image formats. Frames have a `delay` property. Maybe we can give the `Delay` as a property to the type `ImageFrame` instead of using metadata? (Or create a property in the base type `ImageFrameMetadata`)

Ah no, think of `ImageFrame` as merely an in-memory representation of the pixel data with no superflous properties. A `Delay` property would not make sense for an image containing a single frame. 

We've deliberately kept any metadata separate to highlight this and to allow unlimited metadata variations.

**Comment by JimBobSquarePants** [comment]
I need to investigate what is wrong with the encoder. There's two tests failing.

**Comment by Poker-sang** [comment]
> Ah no, think of `ImageFrame` as merely an in-memory representation of the pixel data with no superflous properties. A `Delay` property would not make sense for an image containing a single frame. 
> 
> We've deliberately kept any metadata separate to highlight this and to allow unlimited metadata variations. 

If we convert gif to apng, webp or other animated image formats, the delay information may be lost. Or we need to write appropriate logic for each metadata. 🤔

**Comment by JimBobSquarePants** [comment]
Yes, you would also lose any disposal or blend mode information

**Comment by Poker-sang** [comment]
Ok I see. Thanks for explaining.

**Comment by JimBobSquarePants** [comment]
@Poker-sang I've added support for alpha blending and disposal.

I think it's correct based on how we work with gif/webp but I could do with a second pair of eyes. We desperately need to add test images now to validatethe implementation so if you could add that, it would be great!

**Comment by JimBobSquarePants** [comment]
@Poker-sang Can you give me collaboration access to your fork? There's an open issue with Git LFS which prevents me pushing images to your fork without those permissions.

https://github.com/git-lfs/git-lfs/issues/5001

**Comment by Poker-sang** [comment]
Sure. When I'm cloning a repository lately, filtering content always takes me a long time. Is it caused by too many test images?

**Comment by JimBobSquarePants** [comment]
> Sure. When I'm cloning this repo lately, `Filtering Content` always takes me a long time. Is it caused by too many test images?

I'm not sure what you mean by `Filtering Content`.

**Comment by JimBobSquarePants** [comment]
Ah yeah, there's a lot of LFS files. 

You should only be cloning the repo 1 time really though and using feature branches for your PRs. Every time someone with a fork clones the repo it actually uses up my LFS bandwidth and ends up costing me money if I go over my pre-purchased allowance.

**Comment by Poker-sang** [comment]
Oh I see. I'll pay attention to it.

**Comment by JimBobSquarePants** [comment]
I’ve started adding the tests locally.

**Comment by Poker-sang** [comment]
APNG Specification says:
> If the default image is the first frame:
> ```
> Sequence number    Chunk
> (none)             `acTL`
> 0                  `fcTL` first frame
> (none)             `IDAT` first frame / default image
> 1                  `fcTL` second frame
> 2                  first `fdAT` for second frame
> 3                  second `fdAT` for second frame
> ....
> ```
> If the default image is not part of the animation:
> ```
> Sequence number    Chunk
> (none)             `acTL`
> (none)             `IDAT` default image
> 0                  `fcTL` first frame
> 1                  first `fdAT` for first frame
> 2                  second `fdAT` for first frame
> ....
> ```
Seems like a simple and not very useful feature, do we need to implement it? Says adding an option in the `Encoder` to exclude the first frame.

**Comment by JimBobSquarePants** [comment]
I think that just means the first frame control is optional. The code I have locally handles that.

**Comment by Poker-sang** [comment]
I am so happy to hear that. I am so grateful that you have helped me so much!♥

File: src/ImageSharp/Formats/Png/Chunks/AnimationControl.cs
Comment (Poker-sang) [reply src/ImageSharp/Formats/Png/Chunks/AnimationControl.cs:21]: Do we have to fix this before merging? Or just resolve conversation?

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/Chunks/AnimationControl.cs:21]: Yeah we need to fix this and others. I planned on doing it today

**Review by JimBobSquarePants — APPROVED** [review]
I'm happy with this now. Thanks @Poker-sang for a great addition to the library!

**Pull request merged**
